### PR TITLE
Changes the *attempts_timeout to be usable in milliseconds with a new records entry

### PR DESCRIPTION
**Pull request by ezelkow1**
This PR adds new records entries for *attempts_timeout_ms

If the corresponding _ms version of a parameter is set, not the default of 0, then it will be used in milliseconds instead of the existing parameter which is in seconds.

This is more of a stop gap solution until we get to the yaml rewrite and can add actual types to records

### CODE REVIEW

**Comment by ezelkow1** [comment]
will fix up lua build errors in a bit, saw these with my 7x build but for some reason not on master, but I see them now

**Comment by ezelkow1** [comment]
Ok, I will go look at the record source stuff

**Comment by SolidWallOfCode** [comment]
If you're willing to do some version archeology, this was used to handle the switch over for session server sharing, when the configuration was changed from an integer to a pair of string enumeration values with [some backwards compatibility logic](https://github.com/apache/trafficserver/blob/bd7b9f0165c86ba3fe68e24a462ee454aafd2039/proxy/http/HttpConfig.cc#L186).

**Comment by ezelkow1** [comment]
So is this the sort of stuff you were expecting?

One thing I noticed while testing this on a 7x build, I can see it correctly see the record changes if I enable an _ms record and do a reload, my debug would switch to indicate it was using the _ms value. However if I removed the _ms setting from records.config I would never see it switch back. So I was unsure if this was a bug, only an issue in 7x, or intentional

**Comment by SolidWallOfCode** [comment]
No, I meant something quite a bit different. [See here](https://github.com/SolidWallOfCode/trafficserver/pull/25/commits/919c29d924b1c80a356964f8199ef7dfbbe7a101) for an example. Fundamentally there should be only one value for each timeout, not split by the units. The two different configuration variables should set that value differently, but set the same value. It's much too expensive to check the source for every transaction and unnecessary, as it changes rarely and if it does change, there is a callback to handle it.

**Comment by SolidWallOfCode** [comment]
Yes, much more like that. I think it's cleaner over all and much faster. Does it make sense to you?

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
Yes, that's quite possible as well. If you look at the implementation there's a point where the enumeration is converted to a pointer to a variable - you can simply have the two enumeration values use the same overridable value in the same way as done here. It's actually easier now with the management converters I added recently - this mechanism makes it so you can return different converters for the two enumeration values which differ by how much they multiply the incoming setting.

**Comment by ezelkow1** [comment]
Are you suggesting doing something similar to the TSHTTpTxnConfigStringSet code, so after it pulls the value do a conversion in TSHttpTxnConfigIntGet based on enum?

**Comment by ezelkow1** [comment]
Not sure if this is what you meant.  This is what I did on our 7x branch, but that also doesnt have the fancy converters available

File: src/traffic_server/InkAPI.cc [hunk 1]
```diff
@@ -8274,6 +8300,30 @@ TSHttpTxnConfigIntGet(TSHttpTxn txnp, TSOverridableConfigKey conf, TSMgmtInt *va
 
   *value = conv->get_int(src);
 
+  /* Set the value based on key for these, we either want sec or msec */
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: I think putting this special code in two places is not the best plan. I think it should be handled via the converters. I will try to provide some example code.

**Comment by ezelkow1** [comment]
This seems to pass checks but Im skeptical of the actual value being used since I saw it change to a non-HR time value in the logging and want to make sure the timeouts are still correct

File: lib/ts/apidefs.h.in [hunk 2]
```diff
@@ -710,7 +710,9 @@ typedef enum {
   TS_CONFIG_HTTP_CONNECT_ATTEMPTS_MAX_RETRIES_DEAD_SERVER,
   TS_CONFIG_HTTP_CONNECT_ATTEMPTS_RR_RETRIES,
   TS_CONFIG_HTTP_CONNECT_ATTEMPTS_TIMEOUT,
+  TS_CONFIG_HTTP_CONNECT_ATTEMPTS_TIMEOUT_MS,
```
Comment (SolidWallOfCode) [inline lib/ts/apidefs.h.in]: These changes in the enum make this incompatible with previous versions. Not a problem per se but does make back porting more perilous.

File: proxy/http/HttpConfig.cc [hunk 3]
```diff
@@ -1065,14 +1140,54 @@ HttpConfig::startup()
...
-  HttpEstablishStaticConfigLongLong(c.oride.parent_connect_timeout, "proxy.config.http.parent_proxy.connect_attempts_timeout");
+
+  RecRegisterConfigUpdateCb("proxy.config.http.parent_proxy.connect_attempts_timeout",
+                            &http_parent_proxy_connect_attempt_timeout_cb, &c);
+  RecRegisterConfigUpdateCb("proxy.config.http.parent_proxy.connect_attempts_timeout_ms",
+                            &http_parent_proxy_connect_attempt_timeout_cb, &c);
+  // Do initial load - use ms if set, otherwise use base if set, otherwise just leave it default.
+  if (!http_config_var_is_default("proxy.config.http.parent_proxy.connect_attempts_timeout_ms")) {
+    if (REC_ERR_OKAY == RecGetRecordInt("proxy.config.http.parent_proxy.connect_attempts_timeout_ms", &n)) {
+      c.oride.parent_connect_timeout = HRTIME_MSECONDS(n);
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConfig.cc]: I wonder if this should call `MillisecondsConverter(&c.oride.parent_connect_timeout, n)` for consistency.

**Comment by SolidWallOfCode** [comment]
Sorry, I should have merged it before the big one, you'll have to rebase now.

**Comment by bryancall** [comment]
I am skeptical that < 1 second connect timeouts are needed.  With keep-alive most of the time you should already have a connection.

It looks like this PR still needs some work and hasn't been updated for a few months.  Closing for now.  Please reopen if there are updates.

**Pull request closed without merging**